feat: Add "Show unread count for" setting.

Lalit3716 opened the pull request:
Added the new feature of letting the user choose whether he/she wants to see unread messages count on the left sidebar for streams.

This setting has three options,
"All Streams" - This will show the unread messages count for all streams.
"Unmuted Stream" - This will be the default option and it will only show the unread messages count for unmuted streams. "No Stream" - This option will not show the unread messages count on any stream.

<!-- Describe your pull request here.-->

Fixes: #24149

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

https://user-images.githubusercontent.com/84276404/215797320-0a7966a6-ebe5-41d4-aa8e-e8e7f5c12c7e.mp4

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Lalit3716 commented:
Hi @alya , This PR is ready now for your review please let me know your suggestions, Thanks! 🙂

alya commented:
@Lalit3716 As noted in the PR template, please include still screenshots, in addition to any videos. I'd like to see screenshots from the left sidebar, both in light and dark theme, with and without hover over the relevant stream. Thanks!

Lalit3716 commented:
@alya Sure I uploaded still images in the PR description, please check them out 🙂 !

alya commented:
Thanks! For the hover behavior, we don't want a tooltip. We just want the unread count to show up like it does in other streams when you hover anywhere on the stream row.

Looks good otherwise.

Lalit3716 commented:
Oh Okay! I thought we meant to show a tooltip 😅.
I will fix this right away!

Lalit3716 commented:
@alya Fixed it now, also attached relevant videos in the PR description please have a look, Thanks!

alya commented:
Thanks! The functionality is inconsistent when I test it. E.g., when I pin a muted stream, the counter reappears, though later I saw it disappear again:
<img width="285" alt="Screen Shot 2023-02-04 at 11 49 14 PM" src="https://user-images.githubusercontent.com/2090066/216807783-53bd450a-e27e-4517-b8e8-09b42ae283bb.png">

I'm also not always getting the counter on hover:

<img width="300" alt="Screen Shot 2023-02-04 at 11 49 40 PM" src="https://user-images.githubusercontent.com/2090066/216807790-7741a95d-5eaf-474d-9fe2-3fed3dd6c6cc.png">

Lalit3716 commented:
@alya Sorry my fault, Fixed it, now everything works as expected when pinning/muting/searching streams with the setting on, please test it now, thanks!
But when working on this I found another bug on the main branch related to unread counts and unmuting streams, I have explained the bug in this
[CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Unread.20Count.20issue.20on.20unmuting.20a.20stream) thread please check it out also! Thanks.

timabbott commented on zerver/models.py:
```diff
@@ -1559,6 +1559,17 @@ class UserBaseSettings(models.Model):
+        UNREAD_COUNT_UNMUTED_STREAMS,
+        UNREAD_COUNT_NO_STREAMS,
+    ]
+    show_unread_counts = models.PositiveSmallIntegerField(default=UNREAD_COUNT_UNMUTED_STREAMS)
```
I think maybe `stream_unreads_display` would be a better name for this; I'd expect `show_unread_counts` to be a boolean, not an enum.

I think we've also been moving in the direction of naming the all-caps variables things like `STREAM_UNREADS_DISPLAY_ALL_STREAMS` -- since that helpfully puts these bundles of values into a bit of a namespace. You can see that with `USER_LIST_STYLE_COMPACT`, just above.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8906,6 +8906,23 @@ paths:
+
+            - 1 - All Streams
+            - 2 - Unmuted Streams
+            - 3 - No Streams
```
Streams => streams in all of these; Zulip uses "Sentence case" for capitalization everywhere.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8906,6 +8906,23 @@ paths:
+        - name: show_unread_counts
+          in: query
+          description: |
+            This is used to configure in which streams does user wants to see unread messages count in left sidebar.
```
This could use a bit of polishing; "Configuration for which streams should be displayed with a numeric unread count in the left sidebar. Streams that do not have an unread count will have a simple dot indicator for whether there are any unreads messages."

timabbott commented on static/js/ui_util.ts:
```diff
@@ -47,6 +47,37 @@ export function update_unread_count_in_dom($unread_count_elem: JQuery, count: nu
+        $unread_count_span.hide();
+        $masked_count_span.show();
+    });
+}
```
The frontend logic here needs refactoring; it feels like you have a hundred semi-duplicate `.save()` / `.hide()` calls here.

I also think you can do all of this with CSS, and it'll be a lot nicer. For example, the following would do the toggle-on-hover behavior more reliably than the `mouseenter/mouseleave` handlers:

```
.narrow-filter .out_of_home_view {
    .masked_unread_count {
        display: inline;
     }
   .unread_count {
        display: none;
    }
}

.narrow-filter:hover .out_of_home_view {
    .masked_unread_count {
        display: none;
     }
   .unread_count {
        display: inline;
    }
}
```

You'll need to set a new outer CSS class when rendering rows to communicate the value of the new display display to the CSS logic -- so probably `out_of_home_view` wouldn't actually appear in the CSS, instead being something like `unread_counts_masked` that would be computed in template rendering.

timabbott commented on static/js/stream_list.js:
```diff
@@ -422,6 +452,16 @@ export function update_streams_sidebar(force_rerender) {
+            unread.stream_has_any_unread_mentions(stream_id),
+        );
+    }
+}
```
Is this preferable to just calling `update_streams_sidebar(true)` from the server events code?

Lalit3716 replied:
Actually `update_stream_sidebar` never calls `update_read_counts` method on stream_rows even with `force_rerender=true`,
hence I made this separate function for exactly that 🙂

timabbott replied:
OK but this function is explicitly document as something not to call in a loop:

```
export function stream_has_any_unread_mentions(stream_id) {                               
    // This function is somewhat inefficient and thus should not be                       
    // called in loops, since runs in O(total unread mentions) time.                      
    const streams_with_mentions = unread_topic_counter.get_streams_with_unread_mentions();
    return streams_with_mentions.has(stream_id);                                          
}                                                                                         
```

I'm not sure what the best solution is, but we should either do a single `get_counts` call or do something else.

Lalit3716 replied:
@timabbott I have changed some logic it is much more efficient now we won't need to call `stream_has_unread_mentions` in a loop! Please check out the latest commit, Thanks!

timabbott commented on static/js/stream_list.js:
```diff
@@ -29,23 +29,51 @@ import * as topic_zoom from "./topic_zoom";
+        show_unreads_for.NO_STREAMS || (show_unreads_for.UNMUTED_STREAMS && sub.is_muted);
+
+    return hide_count;
+}
```
This should I think be done using the coding patterns in the `settings_data` / `settings_config` modules; see `using_dark_theme` for an example of how to do this sort of check nicely. I think a `settings_data.should_mask_unread_count(sub.is_muted)` would be the cleanest API, since we don't want to have to import the types for `sub` objects there and we really just need that boolean to make the determination.

timabbott commented:
Thanks for working on this @Lalit3716! I posted what I hope is a pretty complete review of details you can rework to get this into a nice state. I would expect a pure CSS implementation to also eliminate most risk of future "inconsistent hover behavior" bugs.

alya commented:
@Lalit3716 Please post a comment when you've addressed @timabbott 's feedback above and have manually tested your work, and I'll do another testing pass then. Thanks!

Lalit3716 commented:
Sure I will work on the changes right away.

Lalit3716 commented:
@timabbott I have fixed all those changes but I am now getting no 100% test coverage on node test need some help there other than that every thing looks good to me now except the bug I have reported here [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Unread.20Count.20issue.20on.20unmuting.20a.20stream).

alya commented:
>  but I am now getting no 100% test coverage on node test need some help there

Have you asked for help on  CZO?

Lalit3716 commented:
Asked it now [here](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Recover.20test.20coverage/near/1504835)!

Lalit3716 commented:
@timabbott I have addressed your review and fixed them please take a look again!
@alya This is ready for another manual testing now can you please check it once more, thanks!

Lalit3716 commented:
Should I squash the commits @timabbott here ?

timabbott commented on zerver/migrations/0429_merge_20230211_0759.py:
```diff
@@ -0,0 +1,12 @@
+        ("zerver", "0428_remove_realm_email_address_visibility"),
+    ]
+
+    operations = []
```
You should never create merge migrations. You'll want to delete these and read https://zulip.readthedocs.io/en/latest/subsystems/schema-migrations.html for how to renumber migrations correctly instead.

I know Django's built-in text tells you to merge them, but that's only a thing you want to do if we've already merged both versions! :)

Lalit3716 replied:
Ah alright understood, Django sure made it sound like, the only option I have is to merge migrations lol, anyways fixed the migrations!

timabbott commented:
Yes, you should squash these commits; they aren't atomic in a useful way. It's usually not helpful to have commits that just fix the bugs in a previous commit in that same PR -- they just make it harder to read and understand the series of changes we'd be making to `main`. (To be clear, it may well be worth having preparatory refactoring commits in a PR like this that adjust the structure of the code without functional changes in order to make the PR reviewable, but these aren't that).
Check out the [Zulip commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for more details.

Lalit3716 commented:
@timabbott Cleaned up the commit history and fixed the migrations numbering, please have a look!

alya commented:
The unread count should be shown on hovering anywhere over the stream row. Currently, it's not shown when the mouse is to the right of the count.

alya commented:
Also, the menu dropdown needs to be styled like other dropdowns;
<img width="424" alt="Screen Shot 2023-02-25 at 11 54 30 PM" src="https://user-images.githubusercontent.com/2090066/221398913-f28c2e6b-bc7e-4007-b2ba-ac15596b6fe9.png">

alya commented:
Otherwise looks good!

Lalit3716 commented:
@alya Fixed the issues you mentioned, thanks for reviewing it!

alya commented:
I'll retest when conflicts have been resolved. Thanks!

Lalit3716 commented:
@alya Resolved the conflicts!

alya commented:
Thanks! Some notes:

1. There shouldn't be any changes to the horizontal alignment of unread counts. Things are misaligned:
![Screen Shot 2023-03-02 at 3 34 14 PM](https://user-images.githubusercontent.com/2090066/222588606-c6e54a44-34af-423c-bdea-af9e079657ff.png)

2. Unread counts for topics are getting hidden, which is not correct. As noted in the issue:

> Unread counts for topics within streams will still be shown in the left sidebar and Recent conversations (no changes)

![Screen Shot 2023-03-02 at 3 36 29 PM](https://user-images.githubusercontent.com/2090066/222588749-de41590d-5f5d-4bb3-95d7-c6f416e27126.png)

Please update the PR and do a round of careful manual testing. See https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#manual-testing.

alya commented:
Thanks for the update! Playing with the UX some more, let's just always show the count for the active stream (i.e., once you are viewing topics in the stream). The same should apply to the "more topics" view as well.

Lalit3716 commented:
Hi @alya I agree that we should always show the count when a stream is active but it's something that seems a little tricky to implement with respect to my current implementation of the setting, so I would like some suggestions/advice on how to approach this maybe by @timabbott and hence I have opened a [CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Reliable.20way.20to.20find.20out.20active.20stream) thread for it, Thanks!

Lalit3716 commented:
> Thanks for the update! Playing with the UX some more, let's just always show the count for the active stream (i.e., once you are viewing topics in the stream). The same should apply to the "more topics" view as well.

Hi, @alya regarding this I have figured out a solution that will show unread counts irrespective of the display setting if a stream is active, but currently it respects the setting if we are in a topic view it is due to the fact that we don't have any reliable class/id in CSS which tells us that the stream is active despite being in a topic view because currently we removes `active-filter` class when in a topic view.

alya commented:
Is there a question for me here? I really can't comment on the technical approach.

Lalit3716 commented:
Oh okay, no worries I will try to come up with something then.

Lalit3716 commented:
@alya This is ready for another manual review. I have tested it on my end and everything works as expected now 🙂.

Lalit3716 commented:
@timabbott this is ready for a code review.
- Made a prep commit for a new CSS class.
- Cleaned up some work.

Lalit3716 commented:
@alya Updated the screenshots and videos and also included "When a stream is active" gif!

alya commented:
Nice, looks good to me in manual testing now! @amanagr would you be up for reviewing this one?

alya commented:
There isn't necessarily something to do right now, but just noting here that we should be mindful of how this interacts with the following point from #24243.

> If there are unread messages in an unmuted topic within a muted stream, replace the unread counter for the stream with a counter that just adds up all the unread messages in unmuted topics in the stream. The counter should be shown in regular color (not faded). When there are no unread messages in unmuted topics, we should show the stream counter as we currently do.

I think in this situation, it's probably best to show the unread counter even if the user turned off unread counters for muted streams.

Lalit3716 commented:
#### Updates:

- Rebased the PR with better commit history.
- Reworked the logic for live update of the setting and now it's pure css-based!
- Handled the case when there are unread messages in an unmuted topic within a muted stream, in this case, we will just display the unread count for unmuted topics regardless of setting.

[Video](https://user-images.githubusercontent.com/84276404/234644299-9395c13a-dbe2-4633-8b31-610254810910.mp4)

shameondev commented:
Hey @Lalit3716, I noticed you made some updates to the existing changes. Is there anything specific we should know before the integration review?

timabbott commented on web/src/stream_list.js:
```diff
@@ -595,14 +595,20 @@ export function get_sidebar_stream_topic_info(filter) {
     const info = get_sidebar_stream_topic_info(filter);
 
-    deselect_stream_items();
+    deselect_stream_items(info.stream_id);
```
Can you explain how you ensure that `stream-expanded` is set whenever the stream is open? I worry that this is going to be incorrect for cases like where the user navigates to a topic via a link, reloading a browser tab already open to the topic ... or where we rerender the sidebar in the background (e.g. because a user subscribed to a new stream).

Also, `deselect_stream_items` seems like a bad choice of function name for where to put this logic -- `update_stream_sidebar_for_narrow` would make more sense, or its own independent function.

Lalit3716 replied:
So `deselect_stream_items` is the only place where we were removing `active-filter` for all stream rows even if it was indeed expanded this includes all the scenarios you mentioned.

But I agree adding this logic here in this function does not make a whole lot of sense and the logic itself is a little bit excessive. I have refactored this change to add `stream-expanded` class in `update_stream_sidebar_for_narrow` function irrespective of whether a topic is narrowed inside that stream row or not.

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -137,22 +137,70 @@ li.show-more-topics {
+
+        & .masked_unread_count {
+            margin-right: 19px;
+        }
```
Can you add some comments documenting these magic numbers?

Lalit3716 replied:
So basically we have horizontal padding of 4px in addition to 15px margin for `unread_counts` span applied in `app_components.css`. Therefore to keep `masked_unread_count` span horizontally centered with `unread_counts` span we will need to add those extra 4px of padding applied to `unread_counts` and hence a total margin of `15px + 4px = 19px`.

timabbott commented on zerver/models.py:
```diff
@@ -1689,6 +1702,7 @@ class UserBaseSettings(models.Model):
         send_stream_typing_notifications=bool,
         web_mark_read_on_scroll_policy=int,
         user_list_style=int,
+        stream_unreads_display=int,
```
Have we discussed the field name in "api design"? I think "web_stream_unreads_display_policy" would probably be most appropriate, since this is a web app specific setting, and we've been using `_policy` as a suffix for these integer-format fields.

Lalit3716 replied:
No, we haven't discussed the field name yet, just opened this topic on CZO for it - [Topic Link](https://chat.zulip.org/#narrow/stream/378-api-design/topic/property.20name.20for.20.22show.20unread.20counts.22.20user.20setting.2E).

Although I like the name `web_stream_unreads_display_policy`, I will wait for some feedback from folks on CZO before renaming it.

Lalit3716 commented:
@timabbott Changed the settings field name to be `web_stream_unreads_count_display_policy` and addressed all other comments, this is ready for another review now.

timabbott commented:
I renumbered the migrations and applied this change:
``` diff
diff --git a/web/styles/left_sidebar.css b/web/styles/left_sidebar.css
index 3c3fc439a2..95dccc6b49 100644
--- a/web/styles/left_sidebar.css
+++ b/web/styles/left_sidebar.css
@@ -153,8 +153,11 @@ li.show-more-topics {
         }
 
         & .masked_unread_count {
-            /* We add +4px margin for masked unread count span to make it center aligned with unread_count span which has a padding of 4px horizontally in addition to 15px margin. */
-            margin-right: 19px;
+            /* TODO: In theory this should be 19, to make it center aligned with
+               unread_count span which has a padding of 4px
+               horizontally in addition to 15px margin. But 18px looks
+               better centered. */
+            margin-right: 18px;
         }
 
         &.stream-with-count {
```

With this changes, this seems to work well and I've marked it to merge once CI passes, thanks for all the work on this @Lalit3716! 

Depending on how it goes on chat.zulip.org, we may end up disabling it for further fiddling.

@karlstolley I suspect the CSS part of this (basically just the last commit) may be something you'll want to rewrite as part of general left sidebar work, but I think it's broadly within the theme of the existing implementation, so I didn't think it worth blocking on.

@drrosa can you work on a /help/ PR for this new feature?
